Add hive_metastore.thrift machinery needed for alterTable

djsagain opened the pull request:
This commit adds the minimum machinery needed to be able to call metastore.alterTable(alterTableRequest).  That minimum machinery is:

- Adding int field writeId, number 19, at the end of struct Table.
- Adding structs AlterTableRequest and AlterTableResponse.
- Adding new method alter_table_req(alterTableRequest).

Because files hive_metastore.thrift and fb303.thrift didn't exist in the repo, the first commit in this series adds the 3.1.2 versions of those files.

losipiuk committed:
Import hive_metastore.thrift from Hive 3.1.2

losipiuk committed:
Generate java sources from thrift on build

djsagain committed:
Add hive_metastore.thrift machinery needed for alterTable

This commit adds the minimum machinery needed to be able to
call metastore.alterTable(alterTableRequest).  That minimum
machinery is:
o Adding int field writeId, number 19, at the end of struct Table.
o Adding structs AlterTableRequest and AlterTableResponse.
o Adding new method alter_table_req(alterTableRequest).

findepi commented on src/main/thrift/hive_metastore.thrift:
```diff
@@ -1759,6 +1776,9 @@ service ThriftHiveMetastore extends fb303.FacebookService
   // alter table not only applies to future partitions but also cascade to existing partitions
   void alter_table_with_cascade(1:string dbname, 2:string tbl_name, 3:Table new_tbl, 4:bool cascade)
                        throws (1:InvalidOperationException o1, 2:MetaException o2)
+  AlterTableResponse alter_table_req(1:AlterTableRequest req)
```
Which hive version adds this and why don't we have this yet by means of using hive packages?

djsagain replied:
AlterTableRequest, AlterTableResponse and alter_table_req(alterTableRequest) were added in this commit:

```
commit c8f4984fe934af0fb65282bb6218b3baa6c2ca08
Merge: 93b9cdd694 9c5c9409b5
Author: sergey <sershe@apache.org>
Date:   Fri Jul 13 18:43:39 2018 -0700
```

The version of hive_metastore.thrift we get right now in the hive 3.1.2 release is hive-metastore 2.3.3, as determined by the upgrade-acid module. It's also the version of hive_metastore.thrift you get if you checkout a standard download of hive branch rel/release-3.1.2

I'm not sure why that version is used. I've been chatting with @electrum about this stuff, so perhaps he has a better idea.

AFAICT _any_ write operation using Hive ACID needs to call the alter_table_req method, and must fill in the new writeId field of Table.  I'd be delighted to find that some existing version we use of hive_metastore.thrift has the required additions.

findepi replied:
`c8f4984fe934af0fb65282bb6218b3baa6c2ca08` appears in hive `master` after 3.x line.
Was it backported to Hive 3.x and/or Hive 2.x?

electrum replied:
`alter_table_req` doesn't exist in https://github.com/apache/hive/blob/rel/release-3.1.2/standalone-metastore/src/main/thrift/hive_metastore.thrift

It's used in HDP3, so it seems like we should also use it when implementing writes. Of course, this leads to the question as to how ACID works for open source Hive 3.x. This is difficult to answer since none of us have been able to get it running.

findepi replied:
Thanks for explaining. Makes sense. 

> question as to how ACID works for open source Hive 3.x

It matters only if there is sufficient number of open source Hive 3.x users.
Do we know?

I'd guess HDP 3 may be a more popular version than open source Hive 3.x.
Later, we can handle multiple versions with the "alternative call" trick.

electrum replied:
Based on my experience of trying to run it, I might wager to guess zero users, so it's more of a curiosity than a question of practical concern.

findepi commented on src/main/thrift/hive_metastore.thrift:
```diff
@@ -430,7 +430,8 @@ struct Table {
   17: optional string catName,          // Name of the catalog the table is in
-  18: optional PrincipalType ownerType = PrincipalType.USER // owner type of this table (default to USER for backward compatibility)
+  18: optional PrincipalType ownerType = PrincipalType.USER, // owner type of this table (default to USER for backward compatibility)
+  19: optional i64 writeId=-1
```
Which hive version adds this and why don't we have this yet by means of using hive packages?

djsagain replied:
writeId is added by this commit.  However, in that commit, the name used was txnid:

```
commit 1d46608e89c26ed123e96c3b79ef59b50d2349a6
Author: sergey <sershe@apache.org>
Date:   Mon Jun 18 14:50:31 2018 -0700
```

The name of the attribute was changed to writeId in this commit:

```
commit 97f0513c4c8ff1c251b1bdd1b84bd238557f03b0 (HEAD)
Author: Matt McCline <mmccline@hortonworks.com>
Date:   Thu Oct 4 14:37:21 2018 -0500
```

The version of hive_metastore.thrift we get right now in the hive 3.1.2 release is hive-metastore 2.3.3, as determined by the upgrade-acid module.  It's also the version of hive_metastore.thrift you get if you checkout a standard download of hive branch rel/release-3.1.2

I'm not sure why that version is used.  I've been chatting with @electrum  about this stuff, so perhaps he has a better idea.

losipiuk reviewed (commented):
If we want those changes we need to also regenerate .java classes from updated .thrift file. AFAIK those are not generate on build.

electrum commented on pom.xml:
```diff
@@ -545,6 +545,42 @@
+                <groupId>org.codehaus.mojo</groupId>
+                <executions>
+                    <execution>
+                        <id>Generate sources</id>
```
We can remove the ID since there is only one execution

djsagain replied:
Removed.

electrum commented on pom.xml:
```diff
@@ -545,6 +545,42 @@
+                <artifactId>build-helper-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>add-source</id>
```
Same, can remove

djsagain replied:
Removed.

electrum commented on pom.xml:
```diff
@@ -545,6 +545,42 @@
     <build>
         <plugins>
+            <plugin>
+                <artifactId>exec-maven-plugin</artifactId>
```
Order `groupId` followed by `artifactId`, and add the version
```xml
<groupId>org.codehaus.mojo</groupId>
<artifactId>exec-maven-plugin</artifactId>
<version>3.0.0</version>
```

djsagain replied:
Reordered and version added.

electrum commented on pom.xml:
```diff
@@ -545,6 +545,42 @@
+
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>build-helper-maven-plugin</artifactId>
```
Add version
```xml
<version>3.2.0</version>
```

djsagain replied:
Added.

electrum commented on pom.xml:
```diff
@@ -718,6 +754,7 @@
                                     <artifact>org.apache.hive:hive-standalone-metastore</artifact>
                                     <excludes>
                                         <exclude>package.jdo</exclude>
+                                        <exclude>org/apache/hadoop/hive/metastore/api/*</exclude>
```
Change `*` to `**` so we exclude any subdirectories

djsagain committed:
Add metastore call alter_partitions_req and request/response classes

This request is needed to support Hive ACID insert or update or merge
into a partitioned table.

djsagain committed:
Add writeId and stats fields in Partition

Delete and insert Hive ACID operations must update the writeId
of partitions as part of commit, and our hive_metastore.thrift doesn't
have writeId.  This commit adds that field as seen in the master hive
branch, as well as two optional statistics fields, isStatsCompliant
and colStats.

electrum commented on bin/generate-sources.sh:
```diff
@@ -0,0 +1,15 @@
+
+SOURCE_DIR=src/main/thrift
+TARGET_DIR=target/generated-sources/thrift
+rm -rf ${ROOT_DIR}/${TARGET_DIR}
```
All the `ROOT_DIR` paths need to be quoted in case there are spaces or other special characters
```shell
rm -rf "${ROOT_DIR}/${TARGET_DIR}"
```
Run the `shellcheck` program on the script to see the errors. You can also use IntelliJ with the Shell Script plugin installed.
